docs/specification.rst

```diff
@@ -63,606 +61,407 @@ For example,
-structural type.
-For example, instead of
+Protocols can now be combined in any way, to produce a new type
 
 ::
 
-    from collections.abc import Container, Iterable
-    from typing import Protocol, TypeVar
-
-    T = TypeVar("T")
-
-    class IterableContainer(Iterable[T], Container[T], Protocol):
-        ...
+    from typing import Protocol
 
-    def assert_in(target: T, it: IterableContainer[T]) -> bool:
-        if item not in it:
-            raise AssertionError(f"{target} does not occur in {', '.join(map(str, it))}")
-
-users could drop the ``IterableContainer`` class and instead annotate ``it`` as
-``Iterable[T] & Container[T]``.
-
-Source: https://github.com/python/typing/issues/18
-
-
-Self
-----
-
-PEP-673 introduced ``Self``, a simple and intuitive way to annotate methods that return an instance
-of their class.
-If methods or attributes of intersection types return ``Self``-typed values, they should be
-inferred as intersection types.
-For example,
-
-::
+    class CanFoo(Protocol):
+        def foo(self) -> int:
+            ...
 
-    from typing import Self
+        def bar(self) -> int:
+            ...
 
-    class Sample: ...
+    class CanBar(Protocol):
+        def bar(self) -> str
+            ...
 
-    class Mixin:
-        @property
-        def me(self) -> Self: ...
+    def f1(x: CanFoo) -> int:
+        return x.foo()
 
-    a: Sample & Mixin
-    reveal_type(a.me)  # Sample & Mixin
+    def f2(x: Intersection[CanBar, CanFoo]) -> str
+        return x.bar() + str(x.foo())
 
+In this example, f2 can now take any class that can foo and can bar.
+Previously this would have required defining a third class. Note how
+here the bar method in CanBar takes priority.
 
 TypedDict
 ---------
 
-PEP-589 introduced ``TypedDict``, a way to define precise types for dictionaries with a fixed set
-of keys.
-Multiple ``TypedDict`` types could be merged into a single ``TypedDict`` type through subclassing.
-For example,
+The TypedDict class can now be used to express the type of ``__getitem__``
+more accurately:
 
 ::
 
-    from typing import TypedDict
-
-    class Movie(TypedDict):
-        name: str
-        year: int
+    from typing import TypedDict, cast
 
-    class BookBasedMovie(Movie):
-        based_on: str
+    class DefinedGetAttr(TypedDict):
+        foo: int
+        bar: str
 
-With intersection types, ``TypedDict`` types no longer need to be inherited, and can be combined in
-ad-hoc way::
+    class Test:
+        def __getitem__(self, item: str):
+            match item:
+                case "foo":
+                    return 1
+                case "bar":
+                    return "test"
 
-    class BookBased(TypedDict):
-        based_on: str
+    x = cast(Intersection[DefinedGetAttr, Test], Test())
 
-    BookBasedMovie = Movie & BookBased
+    y = x["foo"] # The type here is now int, and the value is 1
 
+Mixed Protocols and Non-Protocols
+---------------------------------
 
-Type narrowing in control flow
```

Comment:
I think this section is important, since type checkers are likely to refine type information based on control flow.

Can you clarify how ordering works with control flow. For example, given

```
def f(x: Any):
  if isinstance(x, T):
    x  # is the type here Any&T or T&Any?

Reply:
Sure - this is an interesting example!

I would say x is type T - let's modify this example slightly:

```py
def f(x: A):
  if isinstance(x, T):
    x  # is the type here A&T or T&A?
```
For the isinstance line to be valid, T must be a subclass of A. Therefore, T is more specific than A, so it's sufficient to say it's type T. Substitute A with Any and the logic still holds.

Reply:
Sure - this is an interesting example!

(removed and modifying as I realised something)

Yeah this is really spicy. I'm going to discuss it in the context of concrete classes, as it's easier to reason about and the same rules apply:

```py
class A:
    pass

class B:
    pass

def f(x: A):
  if isinstance(x, B):
    x  # is the type here A&B or B&A?
```
Here's the weird part - being an instance of B doesn't necessarily mean all of B's methods are exposed. A could have overwritten some. I actually think the type here is as follows:

`Intersection[A,B] | Intersection[B,A]`

A nice feature of the ordered intersection is that it's expressive enough to also become an unordered intersection where needed. Alternatively this could be treated as is (in VS Code the type here is:
`<subclass of A and B>`

Reply:
It seems to me that the desire for ordered intersections arises from a desire to resolve all intersection types eagerly, and not have intersection types propagate. But I think this is not correct.

The type of `(A & B).foo` does not need ordering to decide whether `A` or `B` "wins" -- the type of `(A & B).foo` is simply `A.foo & B.foo`, always. If `A.foo` and `B.foo` are related as subtypes (this includes "same type"), then this intersection will immediately simplify. If `A.foo` and `B.foo` do not have a subtype relationship, then we cannot simplify the intersection, and it propagates.